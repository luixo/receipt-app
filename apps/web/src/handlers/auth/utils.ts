import { TRPCError } from "@trpc/server";

import { DAY } from "~utils";
import type { AccountsId, SessionsSessionId } from "~web/db/models";
import { generateConfirmEmailEmail } from "~web/email/utils";
import type { UnauthorizedContext } from "~web/handlers/context";
import { getEmailClient } from "~web/providers/email";

// How long a session should last
export const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long session should be intact before we update it
export const SESSION_SHOULD_UPDATE_EVERY = 2 * DAY;

export const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);

export const createAuthorizationSession = async (
	ctx: UnauthorizedContext,
	accountId: AccountsId,
) => {
	const uuid: SessionsSessionId = ctx.getUuid();
	const expirationDate = getExpirationDate();
	await ctx.database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const sendVerificationEmail = async (
	ctx: UnauthorizedContext,
	email: string,
	token: string,
) => {
	if (!ctx.emailOptions.active) {
		throw new TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient(ctx).send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(ctx, token),
		});
	} catch (e) {
		throw new TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				/* c8 ignore next */
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
