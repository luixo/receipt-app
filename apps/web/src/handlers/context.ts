import { type H3Event, createEvent } from "@tanstack/react-start/server";
import type { inferProcedureBuilderResolverOptions } from "@trpc/server";
import type { TRPCRequestInfo } from "@trpc/server/http";
import type { IncomingMessage, ServerResponse } from "node:http";

import type { Database } from "~db/database";
import type { TestContext } from "~tests/backend/utils/test";
import type { authProcedure } from "~web/handlers/trpc";
import type { CacheDbOptions } from "~web/providers/cache-db";
import type { EmailOptions } from "~web/providers/email";
import type { ExchangeRateOptions } from "~web/providers/exchange-rate";
import type { Logger } from "~web/providers/logger";
import type { S3Options } from "~web/providers/s3";

type TestContextPicks = Pick<TestContext, "getSalt" | "getUuid"> & {
	database: Database;
	logger: Logger;
	emailOptions: EmailOptions;
	cacheDbOptions: CacheDbOptions;
	exchangeRateOptions: ExchangeRateOptions;
	s3Options: S3Options;
};

export type NetContext = {
	event: H3Event;
	info: TRPCRequestInfo;
};

export type UnauthorizedContext = NetContext & TestContextPicks;

export type AuthorizedContext = inferProcedureBuilderResolverOptions<
	typeof authProcedure
>["ctx"];

export const createContext = (
	{
		req,
		res,
		info,
	}: { req: IncomingMessage; res: ServerResponse; info: TRPCRequestInfo },
	testContext: TestContextPicks,
): UnauthorizedContext => ({
	info,
	event: createEvent(req, res),
	...testContext,
});
