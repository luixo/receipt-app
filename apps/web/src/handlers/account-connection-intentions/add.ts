import { TRPCError } from "@trpc/server";
import { entries } from "remeda";
import { z } from "zod";

import type { AccountsId } from "~db/models";
import { queueCallFactory } from "~web/handlers/batch";
import type { AuthorizedContext } from "~web/handlers/context";
import { authProcedure } from "~web/handlers/trpc";
import { emailSchema, userIdSchema } from "~web/handlers/validation";

const addConnectionIntentionSchema = z.strictObject({
	userId: userIdSchema,
	email: emailSchema,
});
type ConnectionIntention = z.infer<typeof addConnectionIntentionSchema>;

const getTargetUsers = async (
	ctx: AuthorizedContext,
	intentions: readonly ConnectionIntention[],
) =>
	ctx.database
		.selectFrom("users")
		.where(
			"users.id",
			"in",
			intentions.map((intention) => intention.userId),
		)
		.leftJoin("accounts", (qb) =>
			qb.onRef("accounts.id", "=", "users.connectedAccountId"),
		)
		.select([
			"users.id",
			"users.name",
			"accounts.email",
			"users.ownerAccountId",
		])
		.execute();

const getTargetAccounts = async (
	ctx: AuthorizedContext,
	intentions: readonly ConnectionIntention[],
) =>
	ctx.database
		.selectFrom("accounts")
		.where(
			"accounts.email",
			"in",
			intentions.map((intention) => intention.email.lowercase),
		)
		.leftJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", ctx.auth.accountId),
		)
		.select([
			"accounts.id",
			"accounts.avatarUrl",
			"accounts.email",
			"users.name as userName",
		])
		.execute();

const getDirectIntentions = async (
	ctx: AuthorizedContext,
	targetAccountsPromise: ReturnType<typeof getTargetAccounts>,
	targetUsersPromise: ReturnType<typeof getTargetUsers>,
) => {
	const targetAccounts = await targetAccountsPromise;
	const targetUsers = await targetUsersPromise;
	if (targetAccounts.length === 0 || targetUsers.length === 0) {
		return [];
	}
	return ctx.database
		.selectFrom("accountConnectionsIntentions")
		.where("accountConnectionsIntentions.accountId", "=", ctx.auth.accountId)
		.where((qb) =>
			qb.or([
				qb(
					"accountConnectionsIntentions.targetAccountId",
					"in",
					targetAccounts.map(({ id }) => id),
				),
				qb(
					"accountConnectionsIntentions.userId",
					"in",
					targetUsers.map(({ id }) => id),
				),
			]),
		)
		.innerJoin("users", (qb) =>
			qb.onRef("users.id", "=", "accountConnectionsIntentions.userId"),
		)
		.select([
			"accountConnectionsIntentions.userId",
			"accountConnectionsIntentions.targetAccountId",
			"users.name",
		])
		.execute();
};

const getViceVersaIntentions = async (
	ctx: AuthorizedContext,
	targetAccountsPromise: ReturnType<typeof getTargetAccounts>,
) => {
	const targetAccounts = await targetAccountsPromise;
	if (targetAccounts.length === 0) {
		return [];
	}
	return ctx.database
		.selectFrom("accountConnectionsIntentions")
		.where(
			"accountConnectionsIntentions.accountId",
			"in",
			targetAccounts.map(({ id }) => id),
		)
		.where(
			"accountConnectionsIntentions.targetAccountId",
			"=",
			ctx.auth.accountId,
		)
		.select([
			"accountConnectionsIntentions.userId",
			"accountConnectionsIntentions.accountId",
		])
		.execute();
};

const getData = async (
	ctx: AuthorizedContext,
	intentions: readonly ConnectionIntention[],
) => {
	const targetAccountsPromise = getTargetAccounts(ctx, intentions);
	const targetUsersPromise = getTargetUsers(ctx, intentions);

	return {
		targetAccounts: await targetAccountsPromise,
		targetUsers: await targetUsersPromise,
		directIntentions: await getDirectIntentions(
			ctx,
			targetAccountsPromise,
			targetUsersPromise,
		),
		viceVersaIntentions: await getViceVersaIntentions(
			ctx,
			targetAccountsPromise,
		),
	};
};

const getIntentionsOrErrors = (
	ctx: AuthorizedContext,
	intentions: readonly ConnectionIntention[],
	{
		targetAccounts,
		targetUsers,
		viceVersaIntentions,
		directIntentions,
	}: Awaited<ReturnType<typeof getData>>,
) =>
	intentions.map((intention) => {
		const targetUser = targetUsers.find((user) => user.id === intention.userId);
		if (!targetUser) {
			return new TRPCError({
				code: "NOT_FOUND",
				message: `User "${intention.userId}" does not exist.`,
			});
		}
		if (targetUser.ownerAccountId !== ctx.auth.accountId) {
			return new TRPCError({
				code: "FORBIDDEN",
				message: `User "${intention.userId}" is not owned by "${ctx.auth.email}".`,
			});
		}
		if (targetUser.email) {
			return new TRPCError({
				code: "CONFLICT",
				message: `User "${intention.userId}" is already connected to account "${targetUser.email}".`,
			});
		}
		const targetAccount = targetAccounts.find(
			(account) => account.email === intention.email.lowercase,
		);
		if (!targetAccount) {
			return new TRPCError({
				code: "NOT_FOUND",
				message: `Account with email "${intention.email.original}" does not exist.`,
			});
		}
		if (targetAccount.userName) {
			return new TRPCError({
				code: "CONFLICT",
				message: `Account with email "${intention.email.original}" is already connected to user "${targetAccount.userName}".`,
			});
		}
		const directIntentionByAccount = directIntentions.find(
			({ targetAccountId }) => targetAccountId === targetAccount.id,
		);
		if (directIntentionByAccount) {
			return new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to "${intention.email.original}" as user "${directIntentionByAccount.name}".`,
			});
		}
		const directIntentionByUser = directIntentions.find(
			({ userId }) => userId === intention.userId,
		);
		if (directIntentionByUser) {
			return new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user "${directIntentionByUser.name}".`,
			});
		}
		const viceVersaIntention = viceVersaIntentions.find(
			({ accountId }) => accountId === targetAccount.id,
		);
		if (viceVersaIntention) {
			return {
				type: "vice-versa" as const,
				targetAccount: {
					id: targetAccount.id,
					email: targetAccount.email,
					avatarUrl: targetAccount.avatarUrl,
				},
				asUser: {
					id: intention.userId,
					name: targetUser.name,
				},
				viceVersaUser: {
					id: viceVersaIntention.userId,
				},
			};
		}
		return {
			type: "direct" as const,
			targetAccount: {
				id: targetAccount.id,
				email: targetAccount.email,
				avatarUrl: targetAccount.avatarUrl,
			},
			asUser: {
				id: intention.userId,
				name: targetUser.name,
			},
		};
	});

type Intention = Exclude<
	ReturnType<typeof getIntentionsOrErrors>[number],
	TRPCError
>;

const insertViceVersaIntentions = async (
	ctx: AuthorizedContext,
	intentions: Extract<Intention, { type: "vice-versa" }>[],
) => {
	if (intentions.length === 0) {
		return;
	}
	await ctx.database.transaction().execute((tx) =>
		Promise.all([
			...intentions.map((intention) =>
				tx
					.updateTable("users")
					.set({ connectedAccountId: ctx.auth.accountId })
					.where((eb) =>
						eb.and({
							ownerAccountId: intention.targetAccount.id,
							id: intention.viceVersaUser.id,
						}),
					)
					.executeTakeFirst(),
			),
			...intentions.map((intention) =>
				tx
					.updateTable("users")
					.set({ connectedAccountId: intention.targetAccount.id })
					.where((eb) =>
						eb.and({
							ownerAccountId: ctx.auth.accountId,
							id: intention.asUser.id,
						}),
					)
					.executeTakeFirst(),
			),
			tx
				.deleteFrom("accountConnectionsIntentions")
				.where(
					"accountConnectionsIntentions.targetAccountId",
					"=",
					ctx.auth.accountId,
				)
				.where(
					"accountConnectionsIntentions.accountId",
					"in",
					intentions.map((intention) => intention.targetAccount.id),
				)
				.executeTakeFirst(),
		]),
	);
};

const insertDirectIntentions = async (
	ctx: AuthorizedContext,
	intentions: Extract<Intention, { type: "direct" }>[],
) => {
	if (intentions.length === 0) {
		return;
	}
	await ctx.database
		.insertInto("accountConnectionsIntentions")
		.values(
			intentions.map((intention) => ({
				accountId: ctx.auth.accountId,
				targetAccountId: intention.targetAccount.id,
				userId: intention.asUser.id,
				createdAt: new Date(),
			})),
		)
		.execute();
};

const getDuplicates = <T, K extends string>(
	array: readonly T[],
	getKey: (item: T) => K,
) =>
	entries(
		array
			.map(getKey)
			.reduce<Partial<Record<string, number>>>(
				(acc, element) => ({ ...acc, [element]: (acc[element] ?? 0) + 1 }),
				{},
			),
	).filter(([, value]) => value !== 1);

type IntentionOutput = {
	account: {
		id: AccountsId;
		email: string;
		avatarUrl?: string;
	};
	connected: boolean;
	user: {
		name: string;
	};
};

export const batchFn =
	(ctx: AuthorizedContext) =>
	async (
		inputs: readonly z.infer<typeof addConnectionIntentionSchema>[],
	): Promise<(IntentionOutput | TRPCError)[]> => {
		const duplicatedEmails = getDuplicates(
			inputs,
			(intention) => intention.email.lowercase,
		);
		if (duplicatedEmails.length !== 0) {
			throw new TRPCError({
				code: "CONFLICT",
				message: `Expected to have unique emails, got repeating emails: ${duplicatedEmails
					.map(([email, count]) => `"${email}" (${count} times)`)
					.join(", ")}.`,
			});
		}
		const duplicatedUserIds = getDuplicates(
			inputs,
			(intention) => intention.userId,
		);
		if (duplicatedUserIds.length !== 0) {
			throw new TRPCError({
				code: "CONFLICT",
				message: `Expected to have unique user ids, got repeating: ${duplicatedUserIds
					.map(([userId, count]) => `"${userId}" (${count} times)`)
					.join(", ")}.`,
			});
		}
		const data = await getData(ctx, inputs);
		const intentionsOrErrors = getIntentionsOrErrors(ctx, inputs, data);
		const intentions = intentionsOrErrors.filter(
			(
				intentionOrError,
			): intentionOrError is Exclude<typeof intentionOrError, TRPCError> =>
				!(intentionOrError instanceof TRPCError),
		);
		const directIntentions = intentions.filter(
			(intention): intention is Extract<typeof intention, { type: "direct" }> =>
				intention.type === "direct",
		);
		const viceVersaIntentions = intentions.filter(
			(
				intention,
			): intention is Extract<typeof intention, { type: "vice-versa" }> =>
				intention.type === "vice-versa",
		);
		await Promise.all([
			insertDirectIntentions(ctx, directIntentions),
			insertViceVersaIntentions(ctx, viceVersaIntentions),
		]);
		return intentionsOrErrors.map((intentionOrError) => {
			if (intentionOrError instanceof TRPCError) {
				return intentionOrError;
			}
			const matchedDirectIntention = directIntentions.find(
				({ targetAccount }) =>
					targetAccount.id === intentionOrError.targetAccount.id,
			);
			const matchedViceVersaIntention = viceVersaIntentions.find(
				({ targetAccount }) =>
					targetAccount.id === intentionOrError.targetAccount.id,
			);
			/* c8 ignore start */
			if (!matchedViceVersaIntention && !matchedDirectIntention) {
				return new TRPCError({
					code: "INTERNAL_SERVER_ERROR",
					message: `Expected to have an intention in either direct or vice-versa connection list.`,
				});
			}
			/* c8 ignore stop */
			return {
				account: {
					id: intentionOrError.targetAccount.id,
					email: intentionOrError.targetAccount.email,
					avatarUrl: intentionOrError.targetAccount.avatarUrl || undefined,
				},
				connected: Boolean(matchedViceVersaIntention),
				user: {
					name: intentionOrError.asUser.name,
				},
			};
		});
	};

const queueAddConnectionIntenion = queueCallFactory<
	AuthorizedContext,
	z.infer<typeof addConnectionIntentionSchema>,
	IntentionOutput
>(batchFn);

export const procedure = authProcedure
	.input(addConnectionIntentionSchema)
	.mutation(queueAddConnectionIntenion);
