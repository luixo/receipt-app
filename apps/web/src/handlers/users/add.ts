import { TRPCError } from "@trpc/server";
import { isNonNullish } from "remeda";
import { z } from "zod";

import { userNameSchema } from "~app/utils/validation";
import type { UsersId, UsersInitializer } from "~db/models";
import { batchFn as addAccountConnectionIntentions } from "~web/handlers/account-connection-intentions/add";
import { queueCallFactory } from "~web/handlers/batch";
import type { AuthorizedContext } from "~web/handlers/context";
import { authProcedure } from "~web/handlers/trpc";
import { emailSchema } from "~web/handlers/validation";

const addUserSchema = z.strictObject({
	name: userNameSchema,
	publicName: userNameSchema.optional(),
	email: emailSchema.optional(),
});

const getUsers = (
	ctx: AuthorizedContext,
	inputs: readonly z.infer<typeof addUserSchema>[],
) =>
	inputs.map((input) => {
		const id = ctx.getUuid() as UsersId;
		return {
			user: {
				id,
				ownerAccountId: ctx.auth.accountId,
				name: input.name,
				publicName: input.publicName,
			},
			connection: input.email
				? {
						name: input.name,
						email: input.email,
						id,
				  }
				: undefined,
		};
	});

const insertConnections = async (
	ctx: AuthorizedContext,
	connections: ReturnType<typeof getUsers>[number]["connection"][],
) => {
	const nonEmptyConnections = connections.filter(isNonNullish);
	if (nonEmptyConnections.length === 0) {
		return [];
	}
	const intentions = await addAccountConnectionIntentions(ctx)(
		nonEmptyConnections.map((input) => ({
			email: input.email,
			userId: input.id,
		})),
	);
	return connections.map((connection) => {
		if (!connection) {
			return;
		}
		const matchedNonEmptyConnectionIndex = nonEmptyConnections.findIndex(
			(nonEmptyConnection) => nonEmptyConnection === connection,
		);
		const matchedIntentionOrError = intentions[matchedNonEmptyConnectionIndex];
		if (matchedIntentionOrError instanceof TRPCError) {
			throw matchedIntentionOrError;
		}
		return matchedIntentionOrError;
	});
};

const insertUsers = async (
	ctx: AuthorizedContext,
	users: UsersInitializer[],
) => {
	await ctx.database.insertInto("users").values(users).execute();
};

const queueAddUser = queueCallFactory<
	AuthorizedContext,
	z.infer<typeof addUserSchema>,
	{
		id: UsersId;
		connection?: Exclude<
			Awaited<
				ReturnType<ReturnType<typeof addAccountConnectionIntentions>>
			>[number],
			TRPCError
		>;
	}
>((ctx) => async (inputs) => {
	const usersWithConnections = getUsers(ctx, inputs);
	await insertUsers(
		ctx,
		usersWithConnections.map(({ user }) => user),
	);
	const connections = await insertConnections(
		ctx,
		usersWithConnections.map(({ connection }) => connection),
	);
	return usersWithConnections.map(({ user }, index) => ({
		id: user.id,
		connection: connections[index],
	}));
});

export const procedure = authProcedure
	.input(addUserSchema)
	.mutation(queueAddUser);
