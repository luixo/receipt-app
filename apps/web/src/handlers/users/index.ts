import { procedure as add } from "./add";
import { procedure as get } from "./get";
import { procedure as getForeign } from "./get-foreign";
import { procedure as getPaged } from "./get-paged";
import { procedure as remove } from "./remove";
import { procedure as suggest } from "./suggest";
import { procedure as suggestTop } from "./suggest-top";
import { procedure as unlink } from "./unlink";
import { procedure as update } from "./update";

export const router = {
	get,
	getForeign,
	getPaged,
	add,
	remove,
	update,
	unlink,
	suggest,
	suggestTop,
};
