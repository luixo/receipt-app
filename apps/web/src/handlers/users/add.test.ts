import { faker } from "@faker-js/faker";
import { TRPCError } from "@trpc/server";
import { describe, expect } from "vitest";

import {
	MAX_USERNAME_LENGTH,
	MIN_USERNAME_LENGTH,
	userIdSchema,
} from "~app/utils/validation";
import { createAuthContext } from "~tests/backend/utils/context";
import {
	insertAccount,
	insertAccountConnectionIntention,
	insertAccountWithSession,
	insertConnectedUsers,
	insertUser,
} from "~tests/backend/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
	expectUnauthorizedError,
} from "~tests/backend/utils/expect";
import { test } from "~tests/backend/utils/test";
import { t } from "~web/handlers/trpc";
import { runInBand } from "~web/handlers/utils.test";

import { procedure } from "./add";

const createCaller = t.createCallerFactory(t.router({ procedure }));

describe("users.add", () => {
	describe("input verification", () => {
		expectUnauthorizedError((context) =>
			createCaller(context).procedure({ name: faker.person.fullName() }),
		);

		describe("email", () => {
			test("invalid", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);
				const caller = createCaller(await createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							name: faker.person.fullName(),
							email: "invalid@@mail.org",
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "email": Invalid email address`,
				);
			});
		});

		(["name", "publicName"] as const).forEach((field) => {
			describe(field, () => {
				test("minimal length", async ({ ctx }) => {
					const { sessionId } = await insertAccountWithSession(ctx);
					const caller = createCaller(await createAuthContext(ctx, sessionId));
					await expectTRPCError(
						() =>
							caller.procedure({
								name: faker.person.fullName(),
								[field]: "a".repeat(MIN_USERNAME_LENGTH - 1),
							}),
						"BAD_REQUEST",
						`Zod error\n\nAt "${field}": Minimal length for user name is ${MIN_USERNAME_LENGTH}`,
					);
				});

				test("maximum length", async ({ ctx }) => {
					const { sessionId } = await insertAccountWithSession(ctx);
					const caller = createCaller(await createAuthContext(ctx, sessionId));
					await expectTRPCError(
						() =>
							caller.procedure({
								name: faker.person.fullName(),
								[field]: "a".repeat(MAX_USERNAME_LENGTH + 1),
							}),
						"BAD_REQUEST",
						`Zod error\n\nAt "${field}": Maximum length for user name is ${MAX_USERNAME_LENGTH}`,
					);
				});
			});
		});

		test("target email is not registered", async ({ ctx }) => {
			const { sessionId } = await insertAccountWithSession(ctx);
			const caller = createCaller(await createAuthContext(ctx, sessionId));
			const fakeEmail = "non-existent@mail.org";
			await expectTRPCError(
				() =>
					caller.procedure({
						name: faker.person.fullName(),
						email: fakeEmail,
					}),
				"NOT_FOUND",
				`Account with email "${fakeEmail}" does not exist.`,
			);
		});

		describe("email connection intention exceptions", () => {
			test("target email is already connected as another user", async ({
				ctx,
			}) => {
				// Foreign account
				const { id: otherAccountId, email: otherEmail } = await insertAccount(
					ctx,
				);
				// Self account
				const { accountId, sessionId } = await insertAccountWithSession(ctx);
				const [{ name: userName }] = await insertConnectedUsers(ctx, [
					accountId,
					otherAccountId,
				]);
				const caller = createCaller(await createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							name: faker.person.fullName(),
							email: otherEmail,
						}),
					"CONFLICT",
					`Account with email "${otherEmail}" is already connected to user "${userName}".`,
				);
			});

			test("account intention already exists", async ({ ctx }) => {
				// Foreign account
				const { id: otherAccountId, email: otherEmail } = await insertAccount(
					ctx,
				);
				// Self account
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { id: userId, name: userName } = await insertUser(ctx, accountId);
				// Self account's intention to connect to foreign account
				await insertAccountConnectionIntention(
					ctx,
					accountId,
					otherAccountId,
					userId,
				);

				const caller = createCaller(await createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							name: faker.person.fullName(),
							email: otherEmail,
						}),
					"CONFLICT",
					`You already has intention to connect to "${otherEmail}" as user "${userName}".`,
				);
			});
		});

		test("mixed success and fail", async ({ ctx }) => {
			const { sessionId } = await insertAccountWithSession(ctx);
			const caller = createCaller(await createAuthContext(ctx, sessionId));

			const results = await expectDatabaseDiffSnapshot(ctx, () =>
				runInBand([
					() => caller.procedure({ name: faker.person.fullName() }),
					() =>
						caller
							.procedure({
								name: faker.person.fullName(),
								email: "invalid@@mail.org",
							})
							.catch((e) => e),
				]),
			);

			expect(userIdSchema.safeParse(results[0].id).success).toBe(true);
			expect(results[0]).toStrictEqual<(typeof results)[0]>({
				id: results[0].id,
				connection: undefined,
			});
			expect(results[1]).toBeInstanceOf(TRPCError);
		});
	});

	describe("functionality", () => {
		test("user is added - no public name, no email", async ({ ctx }) => {
			// Verifying other accounts are not affected
			await insertAccount(ctx);

			const { sessionId } = await insertAccountWithSession(ctx);
			const caller = createCaller(await createAuthContext(ctx, sessionId));

			const result = await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure({ name: faker.person.fullName() }),
			);
			expect(userIdSchema.safeParse(result.id).success).toBe(true);
			expect(result).toStrictEqual<typeof result>({
				id: result.id,
				connection: undefined,
			});
		});

		test("user is added - public name, no email", async ({ ctx }) => {
			// Verifying other accounts are not affected
			await insertAccount(ctx);

			const { sessionId } = await insertAccountWithSession(ctx);
			const caller = createCaller(await createAuthContext(ctx, sessionId));

			const result = await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure({
					name: faker.person.fullName(),
					publicName: faker.person.fullName(),
				}),
			);
			expect(userIdSchema.safeParse(result.id).success).toBe(true);
			expect(result).toStrictEqual<typeof result>({
				id: result.id,
				connection: undefined,
			});
		});

		describe("user is added - with email", () => {
			test("has a vice versa intention", async ({ ctx }) => {
				// Foreign account
				const {
					id: otherAccountId,
					email: otherEmail,
					avatarUrl: otherAvatarUrl,
				} = await insertAccount(ctx);
				const { id: otherUserId } = await insertUser(ctx, otherAccountId);
				// Self account
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				// Foreign account's intention to connect to self account
				await insertAccountConnectionIntention(
					ctx,
					otherAccountId,
					accountId,
					otherUserId,
				);

				const asName = faker.person.fullName();
				const caller = createCaller(await createAuthContext(ctx, sessionId));
				const result = await expectDatabaseDiffSnapshot(ctx, () =>
					caller.procedure({
						name: asName,
						email: otherEmail,
					}),
				);
				expect(userIdSchema.safeParse(result.id).success).toBe(true);
				expect(result).toStrictEqual<typeof result>({
					connection: {
						connected: true,
						account: {
							id: otherAccountId,
							email: otherEmail,
							avatarUrl: otherAvatarUrl,
						},
						user: { name: asName },
					},
					id: result.id,
				});
			});

			test("doesn't have a vice versa intention", async ({ ctx }) => {
				// Foreign account
				const { email: otherEmail, id: otherAccountId } = await insertAccount(
					ctx,
					{ avatarUrl: null },
				);
				// Self account
				const { sessionId } = await insertAccountWithSession(ctx);

				const asName = faker.person.fullName();
				const caller = createCaller(await createAuthContext(ctx, sessionId));
				const result = await expectDatabaseDiffSnapshot(ctx, () =>
					caller.procedure({
						name: asName,
						email: otherEmail,
					}),
				);
				expect(userIdSchema.safeParse(result.id).success).toBe(true);
				expect(result).toStrictEqual<typeof result>({
					connection: {
						connected: false,
						account: {
							id: otherAccountId,
							email: otherEmail,
							avatarUrl: undefined,
						},
						user: { name: asName },
					},
					id: result.id,
				});
			});
		});

		test("multiple users added", async ({ ctx }) => {
			// Self account
			const { sessionId, accountId } = await insertAccountWithSession(ctx);

			// Direct intention foreign account
			const { email: anotherEmail, id: anotherAccountId } = await insertAccount(
				ctx,
				{ avatarUrl: null },
			);

			// Vice versa intention foreign account
			const { id: otherAccountId, email: otherEmail } = await insertAccount(
				ctx,
				{ avatarUrl: null },
			);
			const { id: otherUserId } = await insertUser(ctx, otherAccountId);
			// Foreign account's intention to connect to self account
			await insertAccountConnectionIntention(
				ctx,
				otherAccountId,
				accountId,
				otherUserId,
			);

			const asName = faker.person.fullName();
			const anotherAsName = faker.person.fullName();

			const caller = createCaller(await createAuthContext(ctx, sessionId));
			const results = await runInBand([
				() => caller.procedure({ name: faker.person.fullName() }),
				() => caller.procedure({ name: asName, email: otherEmail }),
				() => caller.procedure({ name: anotherAsName, email: anotherEmail }),
			]);
			results.forEach((result) => {
				expect(userIdSchema.safeParse(result.id).success).toBe(true);
			});
			expect(results).toStrictEqual<typeof results>([
				{ id: results[0].id, connection: undefined },
				{
					connection: {
						connected: true,
						account: {
							id: otherAccountId,
							email: otherEmail,
							avatarUrl: undefined,
						},
						user: { name: asName },
					},
					id: results[1].id,
				},
				{
					connection: {
						connected: false,
						account: {
							id: anotherAccountId,
							email: anotherEmail,
							avatarUrl: undefined,
						},
						user: { name: anotherAsName },
					},
					id: results[2].id,
				},
			]);
		});
	});
});
