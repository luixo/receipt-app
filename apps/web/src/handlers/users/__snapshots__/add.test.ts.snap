// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`users.add > functionality > user is added - no public name, no email 1`] = `
{
  "added": {
    "users": {
      "id:5cef6eeb-d06c-414e-813b-705457dfa76a": {
        "acceptReceipts": false,
        "connectedAccountId": null,
        "createdAt": 2020-01-01T00:00:00.000Z,
        "exposeReceipts": false,
        "id": "5cef6eeb-d06c-414e-813b-705457dfa76a",
        "name": "Kristopher Kulas",
        "ownerAccountId": "7ee13f62-dbf2-4296-bb95-8154ae2c093a",
        "publicName": null,
        "updatedAt": 2020-01-01T00:00:00.000Z,
      },
    },
  },
}
`;

exports[`users.add > functionality > user is added - public name, no email 1`] = `
{
  "added": {
    "users": {
      "id:3ec84632-ed0b-423d-8191-31544ac6bfc0": {
        "acceptReceipts": false,
        "connectedAccountId": null,
        "createdAt": 2020-01-01T00:00:00.000Z,
        "exposeReceipts": false,
        "id": "3ec84632-ed0b-423d-8191-31544ac6bfc0",
        "name": "Shelia Hessel-Kutch",
        "ownerAccountId": "165535c3-68d2-4e19-b854-8d166ce9c56f",
        "publicName": "Ellen Howell-Mosciski",
        "updatedAt": 2020-01-01T00:00:00.000Z,
      },
    },
  },
}
`;

exports[`users.add > functionality > user is added - with email > doesn't have a vice versa intention 1`] = `
{
  "added": {
    "accountConnectionsIntentions": {
      "accountId:29c912b8-219f-426a-b3d2-67fca04d4210|targetAccountId:e137d89d-f6d2-4cb1-829e-fa58ed8d7ec1": {
        "accountId": "29c912b8-219f-426a-b3d2-67fca04d4210",
        "createdAt": 2020-01-01T00:00:00.000Z,
        "targetAccountId": "e137d89d-f6d2-4cb1-829e-fa58ed8d7ec1",
        "updatedAt": 2020-01-01T00:00:00.000Z,
        "userId": "0fb185c3-c540-4168-a8bf-956c8ce3cbb5",
      },
    },
    "users": {
      "id:0fb185c3-c540-4168-a8bf-956c8ce3cbb5": {
        "acceptReceipts": false,
        "connectedAccountId": null,
        "createdAt": 2020-01-01T00:00:00.000Z,
        "exposeReceipts": false,
        "id": "0fb185c3-c540-4168-a8bf-956c8ce3cbb5",
        "name": "Mr. Jim VonRueden",
        "ownerAccountId": "29c912b8-219f-426a-b3d2-67fca04d4210",
        "publicName": null,
        "updatedAt": 2020-01-01T00:00:00.000Z,
      },
    },
  },
}
`;

exports[`users.add > functionality > user is added - with email > has a vice versa intention 1`] = `
{
  "added": {
    "users": {
      "id:a51d7aae-0076-49df-8284-1a59ea6590e3": {
        "acceptReceipts": false,
        "connectedAccountId": "2344445d-bb11-4058-b0f9-10ed5ddba150",
        "createdAt": 2020-01-01T00:00:00.000Z,
        "exposeReceipts": false,
        "id": "a51d7aae-0076-49df-8284-1a59ea6590e3",
        "name": "Andy O'Reilly",
        "ownerAccountId": "d20a5a4f-9477-4707-8b24-0470962cd596",
        "publicName": null,
        "updatedAt": 2020-01-01T00:01:00.000Z,
      },
    },
  },
  "deleted": {
    "accountConnectionsIntentions": {
      "accountId:2344445d-bb11-4058-b0f9-10ed5ddba150|targetAccountId:d20a5a4f-9477-4707-8b24-0470962cd596": undefined,
    },
  },
  "updated": {
    "users": {
      "id:9b2513a6-420a-4961-8f9c-b2ab9d115d3f": {
        "connectedAccountId": "d20a5a4f-9477-4707-8b24-0470962cd596",
        "updatedAt": 2020-01-01T00:01:00.000Z,
      },
    },
  },
}
`;

exports[`users.add > input verification > mixed success and fail 1`] = `
{
  "added": {
    "users": {
      "id:beb192ca-6815-4215-ac1f-ca9aa502c53b": {
        "acceptReceipts": false,
        "connectedAccountId": null,
        "createdAt": 2020-01-01T00:00:00.000Z,
        "exposeReceipts": false,
        "id": "beb192ca-6815-4215-ac1f-ca9aa502c53b",
        "name": "Rosemary Schiller",
        "ownerAccountId": "f69c89cc-b28f-49ab-8c7b-635f1f9b819c",
        "publicName": null,
        "updatedAt": 2020-01-01T00:00:00.000Z,
      },
    },
  },
}
`;
